### DISCUSSION TITLE
[iOS] Application gets started (headless) on notification reception which causes notification press issues

### DISCUSSION THREAD

**Discussion by piljac1** [invertase/react-native-firebase#5487] [Q&A] [answered]
Hi guys! I have an issue regarding notifications when the app is closed (swiped out of the tray) on iOS. Basically, the notification is received and shown correctly, but it triggers a headless application process, which leads to a rather important issue. Basically, when you open the notification (click on it and open it) within around a minute and a half following reception, it brings the headless application in foreground (showing the splash screen) for approximatively 15 seconds to 1 minute and then it closes. But, if you wait around 1 minute and a half or more after the notification is received, it works as expected: it boots the app and navigates to the appropriate screen.

What could cause this and how do I work around this? I'm aware that it's probably not directly related to this package, but it's part of its ecosystem, so I figured I would ask here. Let me know if I should post elsewhere!

Thanks a lot for your efforts and your product,

- Jacob

**Comment by mikehardy** [comment, marked as answer]
Hi there!

I have to be pedantic, please forgive me:

"notification" == what? FCM? Can you show the exact JSON, minus the device token? Speculation will lead to poor advice

Nothing will be a https://stackoverflow.com/help/minimal-reproducible-example here

**Reply by piljac1** [reply]
Hi ! Thanks for your reply.

Using FCM which is using APNs to relay on iOS.

JSON payload looks like this:

```json
{
    "to": "/topics/my-topic",
    "notification": {
        "title": "The title",
        "body": "The body",
        "sound": "default"
    },
    "data": {
        "id": 12345,
        "channelId": "my-channel-id",
        "storageKey": "A_STORAGE_KEY",
        "routeName": "ARouteName"
    },
    "content_available": true
}
```

**Reply by piljac1** [reply]
Updated my previous reply with the actual payload structure I have atm.

Additionnal info : everything works fine on Android, but the fact that iOS is starting a background process upon notification reception seems to be problematic.

**Reply by mikehardy** [reply]
that content_available key is suspect, that's for data-only messages. I wonder what happens if you take that out.
Also curious about differential testing with other combos of payload, I wonder what happens if it is notification only?

Repeating but fixing a typo here: Nothing will beat a https://stackoverflow.com/help/minimal-reproducible-example here

I have not experienced this personally - in my project when a notification posts on ios headless and I tap on it, it boots the app (as it should, I tapped the notification) then depending on data (or lack thereof) which I can process in handlers it hits main page or navigates to some in-app location

**Reply by piljac1** [reply]
Sorry, I'm bound to a contract which states I can't share project code. However, I'd be happy to anonymize files you think are pertinent to the best of my capacities.

Otherwise, here's some additional info after some tests: when I take out `content_available`, notifications that are received when the application is closed (swiped out) start working when I press them (opens the app and navigates to the correct screen). However, foreground notifications stop working. With `content_available`, I would use React Native Push Notification's `onNotification` method to receive the notification and create a local notification. However, this method is not hit without `content_available`. I then implemented your `onMessage` handler and I can see the notification is getting caught there. So I'll probably be able to sort something out from there and then re-test Android's behavior. I'll let you know if it worked!

**Reply by piljac1** [reply]
Confirmed that everything works now in both iOS and Android. `onNotification` now only treats notifications that the user interacted with (pressed on) and `onMessage` manages the incoming notifications when the app is in the foreground. The `content_available` key was also removed from the JSON FCM payload.

Thanks a lot for your help!
